src/nvme/mi.c

```diff
@@ -41,18 +41,26 @@ static bool nvme_mi_probe_enabled_default(void)
  */
 nvme_root_t nvme_mi_create_root(FILE *fp, int log_level)
 {
-	struct nvme_root *r = calloc(1, sizeof(*r));
+	struct nvme_root *r;
 
+	r = calloc(1, sizeof(*r));
 	if (!r) {
+		errno = ENOMEM;
 		return NULL;
 	}
-	r->log_level = log_level;
-	r->fp = stderr;
+
+	if (!fp)
+		fp = stderr;
+
+	r->log.fp = fp;
+	r->log.level = log_level;
+	nvme_init_default_logging(fp, log_level, false, false);
```

Comment:
Wait, this would overwrite whatever user has set previously. I understand something like this may be needed to maintain backwards compatibility though.

I believe `nvme_init_default_logging()` should never be called anywhere from the library. For instance, we want to initialize default libnvme logging once on libblockdev startup.

Reply:
I know this is really messy. This is basically in there from the initial attempt to everything working. I was not sure if we want to try to keep as close as possible to the old way. But maybe we should just break it now and the the default  logging API right.

In short, I agree we you, we should drop this hack.